### DISCUSSION TITLE
Model/Instantiation in scene graph

### DISCUSSION THREAD

**Discussion by johnwason** [tesseract-robotics/tesseract#505] [Ideas]
Currently the scene graph is made up of joints and links, with no concept of models or hierarchy. In most modern game engines, the "model" definition is separated from the "instantiation" of the model. This means that the geometry and kinematics of the model is stored in a central location, and each instantiation of the model is just a reference to the model, with its current states, and any deviations from the common model definition, such as simulated damage or texture changes. This has significant advantages in terms of memory management etc, since the common data is not replicated. It also enforces a more logical layout, where different actors are independent in the scene graph and don't have issues like name collisions.

Would it make sense to create a "model" node, that points to a database of sub-graphs? I am not sure how boost graph would handle this, but there is probably a way to emulate the current behavior using clever template programming with the model ideas.

**Comment by Levi-Armstrong** [comment]
I am onboard with the concept. @mpowelson and @marip8 What do you think?

**Reply by mpowelson** [reply]
Yeah that seems like the direction we have been wanting to go since [this discussion](https://github.com/ros-industrial-consortium/tesseract/issues/11). Perhaps USD is worth another look?
